src/main.rs

```diff
@@ -1,238 +1,400 @@
+use vulkano::memory::allocator::{AllocationCreateInfo, MemoryUsage, StandardMemoryAllocator};
+use vulkano::pipeline::{
+    graphics::{
+        input_assembly::InputAssemblyState,
+        render_pass::PipelineRenderingCreateInfo,
+        vertex_input::Vertex,
+        viewport::{Viewport, ViewportState},
+    },
+    GraphicsPipeline,
+};
+use vulkano::render_pass::{LoadOp, StoreOp};
+use vulkano::swapchain::{
+    acquire_next_image, AcquireError, Swapchain, SwapchainCreateInfo, SwapchainCreationError,
+    SwapchainPresentInfo,
+};
+use vulkano::sync::{self, FlushError, GpuFuture};
+use vulkano::{Version, VulkanLibrary};
 use vulkano_win::VkSurfaceBuild;
-use winit::event_loop::EventLoop;
-use winit::window::WindowBuilder;
+use winit::event::{Event, WindowEvent};
+use winit::event_loop::{ControlFlow, EventLoop};
+use winit::window::{Window, WindowBuilder};
 
+// https://docs.rs/vulkano-shaders/
 mod vs {
     vulkano_shaders::shader! {
         ty: "vertex",
-        src: "
+        src: r"
             #version 450
+
             layout(location = 0) in vec2 position;
+
             void main() {
                 gl_Position = vec4(position, 0.0, 1.0);
             }
-        "
+        ",
     }
 }
 
 mod fs {
     vulkano_shaders::shader! {
         ty: "fragment",
-        src: "
+        src: r"
             #version 450
+
             layout(location = 0) out vec4 f_color;
+
             void main() {
-                f_color = vec4(1.0, 0.0, 0.0, 1.0);
+                f_color = vec4(1.0, 0.5, 0.3, 1.0);
             }
-        "
+        ",
     }
 }
 
-#[tokio::main]
-async fn main() -> Result<(), reqwest::Error> {
-    let library = VulkanLibrary::new().expect("no local Vulkan library/DLL");
+fn main() {
+    let library = VulkanLibrary::new().unwrap();
+
     let required_extensions = vulkano_win::required_extensions(&library);
+
     let instance = Instance::new(
         library,
         InstanceCreateInfo {
-            enumerate_portability: true,
             enabled_extensions: required_extensions,
+            enumerate_portability: true,
             ..Default::default()
         },
     )
-    .expect("failed to create instance");
-
-    for physical_device in instance.enumerate_physical_devices().unwrap() {
-        println!(
-            "Available device: {}",
-            physical_device.properties().device_name
-        );
-    }
+    .unwrap();
 
     let event_loop = EventLoop::new();
-
-    // Had to downgrade winit to 0.27.5 to get this to work
+    // This presents the surface of the object
     let surface = WindowBuilder::new()
         .build_vk_surface(&event_loop, instance.clone())
         .unwrap();
 
-    use winit::event::{Event, WindowEvent};
-    use winit::event_loop::ControlFlow;
-
-    use vulkano::device::DeviceExtensions;
-
-    let device_extensions = DeviceExtensions {
+    // Swapchain is needed to present images to a surface. The `khr_swapchain` provides that.
```

Comment:
I guess we could just draw on the surface directly while the swapchain gives us the opportunity to be 100% sure that all rendering finished before presenting?

Reply:
This is just me trying to understand things, not any suggestion of how things should be done 😅